cms/djangoapps/contentstore/views/course.py

```diff
@@ -1407,41 +1411,48 @@ def advanced_settings_handler(request, course_key_string):
+                        update_course_advanced_settings(course_module, request.json, request.user)
                     )
+                except ValidationError as err:
+                    return JsonResponseBadRequest(err.detail)
 
-                    if is_valid:
-                        try:
-                            # update the course tabs if required by any setting changes
-                            _refresh_course_tabs(request, course_module)
-                        except InvalidTabsException as err:
-                            log.exception(str(err))
-                            response_message = [
-                                {
-                                    'message': _('An error occurred while trying to save your tabs'),
-                                    'model': {'display_name': _('Tabs Exception')}
-                                }
-                            ]
-                            return JsonResponseBadRequest(response_message)
-
-                        # now update mongo
-                        modulestore().update_item(course_module, request.user.id)
-
-                        return JsonResponse(updated_data)
-                    else:
-                        return JsonResponseBadRequest(errors)
 
-                # Handle all errors that validation doesn't catch
-                except (TypeError, ValueError, InvalidTabsException) as err:
-                    return HttpResponseBadRequest(
-                        django.utils.html.escape(str(err)),
-                        content_type="text/plain"
-                    )
+def update_course_advanced_settings(course_module, data: Dict, user: User):
+    """Helper function to update course advanced settings from API data."""
+    try:
+        # validate data formats and update the course module.
+        # Note: don't update mongo yet, but wait until after any tabs are changed
+        is_valid, errors, updated_data = CourseMetadata.validate_and_update_from_json(
+            course_module,
+            data,
+            user=user,
+        )
+
+        if not is_valid:
+            raise ValidationError(errors)
+
+        try:
+            # update the course tabs if required by any setting changes
+            _refresh_course_tabs(user, course_module)
+        except InvalidTabsException as err:
+            log.exception(str(err))
+            response_message = [
+                {
+                    'message': _('An error occurred while trying to save your tabs'),
+                    'model': {'display_name': _('Tabs Exception')}
+                }
+            ]
+            return ValidationError(response_message)
+
+        # now update mongo
+        modulestore().update_item(course_module, user.id)
+
+        return updated_data
+
+    # Handle all errors that validation doesn't catch
+    except (TypeError, ValueError, InvalidTabsException) as err:
+        return ValidationError(django.utils.html.escape(str(err)))
```

Comment:
why `django.utils.html.escape` is needed here?

Reply:
It's copied over from the existing code here: https://github.com/edx/edx-platform/pull/27608/files#diff-937adee5129f53ead6b3c7c988d70db9e9e95862f17a6f85a1eea2388d12b461L1441-L1444

I'm not sure of the historical reason for adding this.